**START**

Load in pandas and sqlalchemy's create_engine & text
Then connect to Postgres and create the engine

In [19]:
from sqlalchemy import create_engine, text
import pandas as pd
import collections as coll
import geopandas as gpd
import matplotlib.pyplot as plt
import folium
from folium.plugins import MarkerCluster
database_name = 'scooters'
connection_string = f"postgresql://postgres:postgres@localhost:5433/{database_name}"
engine = create_engine(connection_string)

**EDA**

In [20]:
count_all_rows = '''
(SELECT
    'scooters' table,
    COUNT(*)
FROM scooters)
UNION
(SELECT
    'trips' table,
    COUNT(*)
FROM trips);
'''

with engine.connect() as connection:
    counts = pd.read_sql(text(count_all_rows), con = connection)

counts

table     count
0  scooters  73414043
1     trips    565522

In [21]:
find_nulls_scooters = '''
SELECT *
FROM scooters
WHERE NOT(scooters IS NOT NULL);
'''

with engine.connect() as connection:
    nulls_scooters = pd.read_sql(text(find_nulls_scooters), con = connection)

find_nulls_trips = '''
SELECT *
FROM trips
WHERE NOT(trips IS NOT NULL);
'''

with engine.connect() as connection:
    nulls_trips = pd.read_sql(text(find_nulls_trips), con = connection)

nulls_scooters.info()
nulls_trips.info()

nulls_scooters

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 770 entries, 0 to 769
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   pubdatetime  770 non-null    datetime64[ns]
 1   latitude     770 non-null    float64       
 2   longitude    770 non-null    float64       
 3   sumdid       770 non-null    object        
 4   sumdtype     770 non-null    object        
 5   chargelevel  0 non-null      object        
 6   sumdgroup    770 non-null    object        
 7   costpermin   770 non-null    float64       
 8   companyname  770 non-null    object        
dtypes: datetime64[ns](1), float64(3), object(5)
memory usage: 54.3+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 0 entries
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   pubtimestamp    0 non-null      object
 1   companyname     0 non-null      object
 2   triprecordn

pubdatetime   latitude  longitude  \
0   2019-06-15 23:20:36  36.158512 -86.780570   
1   2019-06-15 23:25:37  36.158560 -86.780545   
2   2019-06-15 23:30:37  36.158646 -86.780541   
3   2019-06-15 23:35:37  36.158682 -86.780519   
4   2019-06-15 23:40:37  36.158683 -86.780522   
..                  ...        ...        ...   
765 2019-06-15 20:15:30  36.158659 -86.780577   
766 2019-06-14 19:14:34  36.150274 -86.813108   
767 2019-06-15 20:10:30  36.158659 -86.780577   
768 2019-06-15 20:20:30  36.158659 -86.780577   
769 2019-06-15 20:25:30  36.158659 -86.780577   

                                           sumdid sumdtype chargelevel  \
0    Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
1    Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
2    Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
3    Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
4    Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
..                                            ...      ...         ...   
765  Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
766  Powered-72a04621-3a01-05f2-3c9b-abae6d7387ce  Powered        None   
767  Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
768  Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
769  Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   

    sumdgroup  costpermin companyname  
0     Scooter        0.15        Bolt  
1     Scooter        0.15        Bolt  
2     Scooter        0.15        Bolt  
3     Scooter        0.15        Bolt  
4     Scooter        0.15        Bolt  
..        ...         ...         ...  
765   Scooter        0.15        Bolt  
766   Scooter        0.15        Bolt  
767   Scooter        0.15        Bolt  
768   Scooter        0.15        Bolt  
769   Scooter        0.15        Bolt  

[770 rows x 9 columns]

While the .info() part of the output is a bit counter-intuitave, it shows that there are 770 null values in the scooters table and none in the trips table.

All of the null values are in the chargelevel column, and looking at the full output, they all belong to Bolt and Spin.

In [ ]:
date_range = '''
(SELECT
    'scooters' table,
    MIN(pubdatetime) begin,
    MAX(pubdatetime) end
FROM scooters)
UNION
(SELECT
    'trips' table,
    MIN(pubtimestamp) begin,
    MAX(pubtimestamp) end
FROM trips);
'''

with engine.connect() as connection:
    dates = pd.read_sql(text(date_range), con = connection)

dates

In [ ]:
aug_first = '''
SELECT *
FROM trips
WHERE enddate > '2019-07-31'
LIMIT 100;
'''

with engine.connect() as connection:
    late = pd.read_sql(text(aug_first), con = connection)

late

Both tables contain 3 months of data and both begin on May 1st, but the trips table ends a day after the scooters table. Looking at why, the trips all began before midnight on July 31st.

In [ ]:
long_trips = '''
SELECT 
    companyname,
    COUNT(*)
FROM trips
WHERE tripduration > 1440
GROUP BY companyname;
'''

with engine.connect() as connection:
    long = pd.read_sql(text(long_trips), con = connection)

long_trips

In [ ]:
short_trips = '''
SELECT
    companyname,
    COUNT(*)
FROM trips
WHERE tripduration < 1
    AND tripdistance <= 0
GROUP BY companyname;
'''

with engine.connect() as connection:
    short = pd.read_sql(text(short_trips), con = connection)

short

This data was supposed to have been cleaned before being submitted to the city, which includes stripping out all trips shorter than a minute or longer than 24 hours.

There are ~7,000 trips longer than 24 hours and over 9,000 trips shorter than one minute.

In [ ]:
available_scooters = '''
SELECT
    companyname company,
    COUNT(DISTINCT sumdid) scooters
FROM scooters
GROUP BY companyname
'''

with engine.connect() as connection:
    available = pd.read_sql(text(available_scooters), con = connection)

active_scooters = '''
SELECT
    companyname company,
    COUNT(DISTINCT sumdid) scooters
FROM trips
GROUP BY companyname
'''

with engine.connect() as connection:
    active = pd.read_sql(text(active_scooters), con = connection)

available

In [ ]:
active

In [ ]:
availability = '''
SELECT
    companyname company,
    COUNT(DISTINCT sumdid) total_scooters
FROM scooters
WHERE sumdid NOT IN
    (SELECT
        DISTINCT sumdid
    FROM trips)
GROUP BY companyname;
'''

with engine.connect() as connection:
    unavailable = pd.read_sql(text(availability), con = connection)
    
unavailable

In [ ]:
daily_use = '''
SELECT 
    DISTINCT sumdid,
    companyname company,
    ROUND(AVG(COUNT(sumdid)) OVER (PARTITION BY sumdid, DATE(pubtimestamp)),2) avg_daily_usage
FROM trips
GROUP BY sumdid, company, pubtimestamp
'''

with engine.connect() as connection:
    usage = pd.read_sql(text(daily_use), con = connection)

multiple_uses = usage[usage['avg_daily_usage'] > 1.00]
multiple_uses

Filter trips table:
- Remove trips under a minute
- Remove all zero-distance trips
- Remove unreasonable long trips
    - These scooters have an average top speed of 15 mph, and generally have a range of about 50 miles.
    - That said, the batteries should last on average ~3.5 hours

In [22]:
trips_clean = '''
SELECT *
FROM trips
WHERE tripduration > 1.0
    AND tripduration < 200.0
    AND tripdistance > 0
    AND tripdistance < 264000
'''

with engine.connect() as connection:
    trips = pd.read_sql(text(trips_clean), con = connection)
    
trips.describe()

trips.head()

pubtimestamp companyname triprecordnum          sumdid  \
0 2019-06-02 20:58:41.033        SPIN        SPI886  Powered3745490   
1 2019-06-02 20:58:41.033        SPIN        SPI877  Powered8106751   
2 2019-06-02 20:58:41.033        SPIN        SPI879  Powered6903219   
3 2019-06-02 20:58:41.033        SPIN        SPI884  Powered6695098   
4 2019-06-02 20:58:41.033        SPIN        SPI863  Powered6228267   

   tripduration  tripdistance   startdate starttime     enddate   endtime  \
0           4.0     793.96328  2019-06-02  20:54:29  2019-06-02  20:57:54   
1          13.0     370.73492  2019-06-02  20:45:48  2019-06-02  20:58:31   
2          13.0     351.04988  2019-06-02  20:46:25  2019-06-02  20:58:37   
3           5.0     131.23360  2019-06-02  20:52:57  2019-06-02  20:57:11   
4          21.0    8674.54096  2019-06-02  20:34:10  2019-06-02  20:54:24   

   startlatitude  startlongitude  endlatitude  endlongitude  \
0      36.161966      -86.775011    36.163276    -86.777170   
1      36.152710      -86.785055    36.152981    -86.783840   
2      36.152711      -86.785046    36.152987    -86.783894   
3      36.161227      -86.777061    36.161449    -86.776705   
4      36.176285      -86.785433    36.152841    -86.790189   

                                           triproute               create_dt  
0  [(36.163276406038705, -86.7771697138313), (36.... 2019-06-03 09:05:38.947  
1  [(36.1529805561506, -86.78383959028197), (36.1... 2019-06-03 09:05:38.980  
2  [(36.152987202313604, -86.78389426748939), (36... 2019-06-03 09:05:39.010  
3  [(36.1614456, -86.7767024), (36.1614472, -86.7... 2019-06-03 09:05:38.880  
4  [(36.15284125996444, -86.79018921571804), (36.... 2019-06-03 09:05:38.753

**MAPPING**

In [23]:
zipcodes = gpd.read_file('../data/zipcodes.geojson')

zipcodes = zipcodes[['zip', 'po_name', 'geometry']]
# Create separate tables for start & end location data
scooter_start = trips[['companyname', 'triprecordnum', 'sumdid', 'startlongitude', 'startlatitude']]
scooter_start_geo = gpd.GeoDataFrame(scooter_start, crs= zipcodes.crs, geometry=gpd.points_from_xy(scooter_start.startlongitude, scooter_start.startlatitude))
scooter_end = trips[['companyname', 'triprecordnum', 'sumdid', 'endlongitude', 'endlatitude']]
scooter_end_geo = gpd.GeoDataFrame(scooter_end, crs= zipcodes.crs, geometry=gpd.points_from_xy(scooter_end.endlongitude, scooter_end.endlatitude))

In [24]:
# Combine start and end locations by zip code
starting_zip = gpd.sjoin(scooter_start_geo, zipcodes, predicate='within')
ending_zip = gpd.sjoin(scooter_end_geo, zipcodes, predicate='within')

In [25]:
starting_zip.head()

companyname triprecordnum          sumdid  startlongitude  startlatitude  \
0         SPIN        SPI886  Powered3745490      -86.775011      36.161966   
3         SPIN        SPI884  Powered6695098      -86.777061      36.161227   
11        SPIN        SPI894  Powered8683450      -86.776591      36.163797   
13        SPIN        SPI888  Powered5330088      -86.774547      36.159697   
14        SPIN        SPI893  Powered8683450      -86.776228      36.163546   

                      geometry  index_right    zip    po_name  
0   POINT (-86.77501 36.16197)           36  37201  NASHVILLE  
3   POINT (-86.77706 36.16123)           36  37201  NASHVILLE  
11  POINT (-86.77659 36.16380)           36  37201  NASHVILLE  
13  POINT (-86.77455 36.15970)           36  37201  NASHVILLE  
14  POINT (-86.77623 36.16355)           36  37201  NASHVILLE

In [26]:
ending_zip.head()

companyname triprecordnum          sumdid  endlongitude  endlatitude  \
0         SPIN        SPI886  Powered3745490    -86.777170    36.163276   
3         SPIN        SPI884  Powered6695098    -86.776705    36.161449   
11        SPIN        SPI894  Powered8683450    -86.776694    36.163699   
12        SPIN        SPI868  Powered5743469    -86.776375    36.160219   
14        SPIN        SPI893  Powered8683450    -86.776350    36.163440   

                      geometry  index_right    zip    po_name  
0   POINT (-86.77717 36.16328)           36  37201  NASHVILLE  
3   POINT (-86.77671 36.16145)           36  37201  NASHVILLE  
11  POINT (-86.77669 36.16370)           36  37201  NASHVILLE  
12  POINT (-86.77637 36.16022)           36  37201  NASHVILLE  
14  POINT (-86.77635 36.16344)           36  37201  NASHVILLE

Now let's look at where scooter trips are taking people. Counting up the number of trips that began in each zip code and comparing them with the number of trips ending in each zip code, we can see which areas people are more likely to come from and go to.

In [27]:
start_by_zip = pd.DataFrame(starting_zip['zip'].value_counts())
end_by_zip = pd.DataFrame(ending_zip['zip'].value_counts())
delta = lambda start_by_zip, end_by_zip : end_by_zip - start_by_zip


In [28]:
destinations = delta(start_by_zip,end_by_zip).sort_values(by='count', ascending=False)
destinations.head(3)

count
zip          
37201  1804.0
37207  1022.0
37209   910.0

In [29]:
origins = delta(start_by_zip,end_by_zip).sort_values(by='count', ascending=True)
origins.head(3)

count
zip          
37204 -4072.0
37219 -2263.0
37203 -1066.0

More trips ended in 37201, 37207, and 37209 than started there, while more trips began in 37204, 37219, and 37203 than ended there. Now let's look into trips to/from these zip areas.

In [34]:
origin_destination_trips= trips[['companyname', 'triprecordnum', 'sumdid', 'startlongitude', 'startlatitude', 'endlongitude', 'endlatitude']]
origin_trips_geo = gpd.GeoDataFrame(origin_destination_trips, crs= zipcodes.crs, geometry=gpd.points_from_xy(origin_destination_trips.endlongitude, origin_destination_trips.endlatitude))
dest_trips_geo = gpd.GeoDataFrame(origin_destination_trips, crs= zipcodes.crs, geometry=gpd.points_from_xy(origin_destination_trips.startlongitude, origin_destination_trips.startlatitude))

In [35]:
origin_zip = gpd.sjoin(origin_trips_geo, zipcodes, predicate='within')
destination_zip = gpd.sjoin(dest_trips_geo, zipcodes, predicate='within')

In [36]:
origin_zip.head()

companyname triprecordnum          sumdid  startlongitude  startlatitude  \
0         SPIN        SPI886  Powered3745490      -86.775011      36.161966   
3         SPIN        SPI884  Powered6695098      -86.777061      36.161227   
11        SPIN        SPI894  Powered8683450      -86.776591      36.163797   
12        SPIN        SPI868  Powered5743469      -86.779857      36.163267   
14        SPIN        SPI893  Powered8683450      -86.776228      36.163546   

    endlongitude  endlatitude                    geometry  index_right    zip  \
0     -86.777170    36.163276  POINT (-86.77717 36.16328)           36  37201   
3     -86.776705    36.161449  POINT (-86.77671 36.16145)           36  37201   
11    -86.776694    36.163699  POINT (-86.77669 36.16370)           36  37201   
12    -86.776375    36.160219  POINT (-86.77637 36.16022)           36  37201   
14    -86.776350    36.163440  POINT (-86.77635 36.16344)           36  37201   

      po_name  
0   NASHVILLE  
3   NASHVILLE  
11  NASHVILLE  
12  NASHVILLE  
14  NASHVILLE

In [37]:
destination_zip.head()

companyname triprecordnum          sumdid  startlongitude  startlatitude  \
0         SPIN        SPI886  Powered3745490      -86.775011      36.161966   
3         SPIN        SPI884  Powered6695098      -86.777061      36.161227   
11        SPIN        SPI894  Powered8683450      -86.776591      36.163797   
13        SPIN        SPI888  Powered5330088      -86.774547      36.159697   
14        SPIN        SPI893  Powered8683450      -86.776228      36.163546   

    endlongitude  endlatitude                    geometry  index_right    zip  \
0     -86.777170    36.163276  POINT (-86.77501 36.16197)           36  37201   
3     -86.776705    36.161449  POINT (-86.77706 36.16123)           36  37201   
11    -86.776694    36.163699  POINT (-86.77659 36.16380)           36  37201   
13    -86.770678    36.155550  POINT (-86.77455 36.15970)           36  37201   
14    -86.776350    36.163440  POINT (-86.77623 36.16355)           36  37201   

      po_name  
0   NASHVILLE  
3   NASHVILLE  
11  NASHVILLE  
13  NASHVILLE  
14  NASHVILLE

In [38]:
# Get centroid of Davidson County
county = gpd.read_file('../data/Davidson County Border (GIS).geojson')
county.geometry.centroid
# Use ID #0
center = county.geometry.centroid[0]
map_center = [center.y, center.x]

# Draw map of starting positions for each trip
scooter_start_map = folium.Map(location = map_center, zoom_start = 10)
# Add marker cluster
marker_cluster = MarkerCluster().add_to(scooter_start_map)
# Add zipcodes to map
folium.GeoJson(zipcodes).add_to(scooter_start_map)
# Use a for loop to add projects
for row_index, row_values in destination_zip.iterrows():
    loc = [row_values['startlatitude'], row_values['startlongitude']]
    pop = str(row_values['companyname'])
    icon=folium.Icon(color="blue",icon="exclamation-triangle", prefix='fa')
    
    marker = folium.Marker(
        location = loc, 
        popup = pop,
   icon = icon) 
    
    marker.add_to(marker_cluster)

scooter_start_map.save('../maps/trip_start.html')
# Display the map
scooter_start_map

C:\Users\chabi\AppData\Local\Temp\ipykernel_18208\207800406.py:3: UserWarning: Geometry is in a geographic CRS. Results from 'centroid' are likely incorrect. Use 'GeoSeries.to_crs()' to re-project geometries to a projected CRS before this operation.

  county.geometry.centroid
C:\Users\chabi\AppData\Local\Temp\ipykernel_18208\207800406.py:5: UserWarning: Geometry is in a geographic CRS. Results from 'centroid' are likely incorrect. Use 'GeoSeries.to_crs()' to re-project geometries to a projected CRS before this operation.

  center = county.geometry.centroid[0]


In [ ]:
# Draw map of ending positions for each trip
scooter_end_map = folium.Map(location = map_center, zoom_start = 10)
# Add marker cluster
marker_cluster = MarkerCluster().add_to(scooter_start_map)
# Add zipcodes to map
folium.GeoJson(zipcodes).add_to(scooter_start_map)
# Use a for loop to add projects
for row_index, row_values in origin_zip.iterrows():
    loc = [row_values['startlatitude'], row_values['startlongitude']]
    pop = str(row_values['companyname'])
    icon=folium.Icon(color="blue",icon="exclamation-triangle", prefix='fa')
    
    marker = folium.Marker(
        location = loc, 
        popup = pop,
   icon = icon) 
    
    marker.add_to(marker_cluster)

scooter_end_map.save('../maps/trip_end.html')
# Display the map
scooter_end_map